README.md

```diff
@@ -1,11 +1,458 @@
+bugs early and prevent their damaging effects.
+
+## Basic C++ Usage
+
+This is a header-only library.  To use it from C++, simply put the
+`include` directory of this repository in your `#include` path, and
+`#include <adobe/contract_checks.hpp>`.
+
+```c++
+#include <adobe/contract_checks.hpp>
+```
+
+Every executable needs a single contract violation handler that
+determines the program's behavior when a violation is detected.  A
+good starting point is provided by a macro that you can expand in a
+source file such as the one containing your `main` function.
+
+```c++
+ADOBE_DEFAULT_CONTRACT_VIOLATION_HANDLER() // no semicolon
+```
+
+There are three primary macros used to check contracts, each with two
+forms: `ADOBE_PRECONDITION`, `ADOBE_POSTCONDITION`, and
+`ADOBE_INVARIANT`. Each has one required argument and one optional
+argument:
+
+- `condition`: an expression convertible to `bool`; if `false`, the
+  violation handler is invoked.
+- `message`: an expression convertible to `const char*` pointing to a
+  [null-terminated](https://en.cppreference.com/w/cpp/string/byte)
+  message that is additionally passed to the violation handler. The
+  default value is the empty string, `""`.
+
+For example,
+
+```c++
+#include <adobe/contract_checks.hpp>
+#include <climits>
+
+// A half-open range of integers.
+// - Invariant: start() <= end().
+class int_range {
+  // The lower bound; if the range is non-empty, its
+  // least contained value.
+  int _start;
+  // The upper bound; if the range is non-empty, one
+  // greater than its greatest contained value.
+  int _end;
+
+  void check_invariant() { ADOBE_INVARIANT(start() <= end()); }
+public:
+  // An instance with the given bounds.
+  // Precondition: end >= start
+  int_range(int start, int end) : _start(start), _end(end) {
+    ADOBE_PRECONDITION(end >= start, "invalid range bounds.");
+    check_invariant();
+  }
+
+  // Returns the lower bound: if *this is non-empty, its
+  // least contained value.
+  int start() const { return _start; }
+
+  // Returns the upper bound; if *this is non-empty, one
+  // greater than its greatest contained value.
+  int end() const { return _end; }
+
+  // Increases the upper bound by 1.
+  // Precondition: end() < INT_MAX
+  void grow_rightward() {
+    ADOBE_PRECONDITION(end() < INT_MAX);
+    int old_end = end();
+    _end += 1;
+    ADOBE_POSTCONDITION(end() == old_end + 1);
+    check_invariant();
+  }
+
+  // more methods...
+};
+```
+
+## CMake Usage
+
+  ```cmake
+  include(FetchContent)
+  if(PROJECT_IS_TOP_LEVEL)
+    FetchContent_Declare(
+      adobe-contract-checks
+      GIT_REPOSITORY https://github.com/stlab/adobe-contract-checks.git
+      GIT_TAG        <this library's release version>
+    )
+    FetchContent_MakeAvailable(adobe-contract-checks)
+  endif()
+  find_package(adobe-contract-checks)
+
+  add_library(my-library my-library.cpp)
+  target_link_libraries(my-library PRIVATE adobe-contract-checks)
```

Comment:
I'm assuming `target_link_libraries` is adding it to the include paths - is there any way to add it to the include paths for all targets instead of repeating these?

Reply:
Yes.  And no, not a principled way.  This is the right way to do it.

Reply:
Even if only some of those targets depend on the library supplying those includes?  I don't think that's principled at all.

Reply:
I'm calling this resolved.  Reopen if you think there's something more to say.